netdata/netdata
# Implement new incremental parser

Fixes #8180
##### Summary
Implements new incremental parser to be used by internal plugins and slaves. 

Usage:
- A new parser is allocated with an appropriate `input` stream
  - An optional `read` and `eof` function can be supplied (defaults to the current implementation
    which expects a `FILE *`
- The pluginss.d language keywords are registered with callback functions
- Enter a loop to
  - Fetch next line from the input stream
  - Call the action function
     - Exit on error
  
##### Component Name
Streaming, Internal collectors

##### Test Plan
- Compile a master agent with the changes
  - Setup a slave that connects to the master without SSL
  - All internal plugins and slave connections should work as before
    - Notice in the logfile message like the following
```
netdata INFO  : PLUGINSD[go.d] : Registered 10 keywords for the parser
netdata INFO  : PLUGINSD[nfacct] : Registered 10 keywords for the parser
netdata INFO  : PLUGINSD[charts.d] : Registered 10 keywords for the parser
netdata INFO  : STREAM_RECEIVER[ubuntu1804,[192.168.122.103]:57304] : Registered 10 keywords for the pars
```
as plugins and incoming slaves initialize a parser and register keywords for processing

Forcing a plugin or a slave to stop will log a `PARSER ended` message as follows:
```
netdata INFO  : PLUGINSD[apps] : PARSER ended
netdata ERROR : PLUGINSD[apps] : '/home/stelios/netdata-inst//netdata/usr/libexec/netdata/plugins.d/apps.plugin' (pid 2982315) disconnected after 9380 successful data collections (ENDs).
netdata ERROR : PLUGINSD[apps] : child pid 2982315 killed by signal 15.
netdata INFO  : PLUGINSD[apps] : '/home/stelios/netdata-inst//netdata/usr/libexec/netdata/plugins.d/apps.plugin' (pid 2982315) was killed with SIGTERM. Disabling it.
netdata INFO  : PLUGINSD[apps] : thread with task id 2982299 finished
netdata INFO  : STREAM_RECEIVER[ubuntu1804,[192.168.122.103]:59308] : PARSER ended
netdata ERROR : STREAM_RECEIVER[ubuntu1804,[192.168.122.103]:59308] : STREAM ubuntu1804 [receive from [192.168.122.103]:59308]: disconnected (completed 4115 updates).
netdata INFO  : STREAM_RECEIVER[ubuntu1804,[192.168.122.103]:59308] : STREAM ubuntu1804 [receive from [192.168.122.103]:59308]: receive thread ended (task id 2982546)
```
- Stop the master and slave
  - Configure slave to connect to the master using SSL
    - Repeat the test as above
##### Additional Information
